clang/lib/CConv/RewriteUtils.cpp

```diff
@@ -360,9 +361,46 @@ std::string ArrayBoundsRewriter::getBoundsString(PVConstraint *PV,
   return BString;
 }
 
+std::set<PersistentSourceLoc *> RewriteConsumer::EmittedDiagnostics;
+void RewriteConsumer::emitRootCauseDiagnostics(ASTContext &Context) {
+  clang::DiagnosticsEngine &DE = Context.getDiagnostics();
+  unsigned ID = DE.getCustomDiagID(DiagnosticsEngine::Warning,
+                                   "Root cause of unchecked pointers: %0");
+  auto I = Info.getInterimConstraintState();
+  SourceManager &SM = Context.getSourceManager();
+  for (auto &WReason : I.RealWildPtrsWithReasons) {
+    if (I.PtrSourceMap.find(WReason.first) != I.PtrSourceMap.end()) {
+      PersistentSourceLoc *PsInfo = I.PtrSourceMap[WReason.first];
+      // Avoid emitting the same diagnostic message twice.
+      if (EmittedDiagnostics.find(PsInfo) == EmittedDiagnostics.end()) {
+        // Convert the PSL into a clang::SourceLocation that can be used with
+        // the DiagnosticsEngine.
+        auto File = SM.getFileManager().getFile(PsInfo->getFileName());
+        SourceLocation SL = SM.translateFileLineCol(File, PsInfo->getLineNo(),
+                                                    PsInfo->getColSNo());
+        // Limit emitted root causes to those that effect more than one pointer
+        // or are in the main file of the TU. Alternatively, don't filter causes
+        // if -warn-all-root-cause is passed.
+        if (WarnAllRootCause || SM.isInMainFile(SL)
```

Comment:
If you are in the main file then you will output the warning even if there's just one wild pointer?

Reply:
In this case main file is per translation unit, so for converting vsftpd, the main files are all the `c` files passed in as command line arguments. The idea is that all wild pointers in these files are interesting whereas a wild pointer in an included header file that doesn't cause wildness anywhere else is generally not interesting.